src/type_object.rs

```diff
@@ -31,16 +33,21 @@ pub trait PySizedLayout<T>: PyLayout<T> + Sized {}
 /// - `Self::AsRefTarget` must have the same layout as `UnsafeCell<ffi::PyAny>`.
 pub unsafe trait HasPyGilRef {
     /// Utility type to make Py::as_ref work.
+    #[cfg(feature = "gil-refs")]
     type AsRefTarget: PyNativeType;
 }
 
+#[cfg(feature = "gil-refs")]
 unsafe impl<T> HasPyGilRef for T
 where
     T: PyNativeType,
 {
     type AsRefTarget = Self;
 }
 
+#[cfg(not(feature = "gil-refs"))]
+unsafe impl<T> HasPyGilRef for T {}
+
```

Comment:
Hmm, is this right? Does it cause any soundness issues if its implemented for everything?

Reply:
The idea was, that implementing it for everything should be functionally equivalent to removing the requirement. There are not APIs left (at least I could find any) that have `HasPyGilRef` as a direct bound without `gil-refs`. So this blanket is only there to satisfy the requirement for `PyTypeCheck` and `PyTypeInfo`, which currently use it as a supertrait.

Reply:
So is this temporary, i.e. will we feature-gate the whole trait and its usage as a super trait later on?

Reply:
I was thinking temporary until we fully remove the `gil-refs` in 0.23, at which point we can also remove the trait and the supertrait requirement.

Reply:
Couldn't we already feature gate the trait (and bounds) in 0.22 based on the feature?

Reply:
I think all the bounds are already gated. We could also gate the whole trait, but then we have to duplicate the definitions of `PyTypeCheck` and `PyTypeInfo` to conditionally include the supertrait. (Or we could could introduce a helper trait in the middle to bridge, but I think that is more or less I've done here, just with a different name)

Reply:
> We could also gate the whole trait, but then we have to duplicate the definitions of PyTypeCheck and PyTypeInfo to conditionally include the supertrait.

I think I would prefer that (we could use a small local helper macro like we had for the const thread_local! syntax. Personally, I would find that less weird than the universal trait impl. I do see any unintended side effects though, so this appears to be purely a matter of preference to me.

Reply:
Sure thing, `PyTypeCheck` and `PyTypeInfo` are actually not _that_ big, so introducing a macro for them is maybe a bit overkill and maybe having them duplicated until 0.22 is out is not such big of a deal.